docs/frog.1
.TH frog 1 "2023 feb 22"

.SH NAME
frog \- Dutch Natural Language Toolkit
.SH SYNOPSIS
frog [\-t] test\-file

frog [options]

.SH DESCRIPTION
Frog is an integration of memory\(hy-based natural language processing (NLP)
modules developed for Dutch.
Frog's current version will (optionally) tokenize, tag, lemmatize, and
morphologically segment word tokens in Dutch text files, add IOB chunks,
add Named Entities and will assign a dependency graph to each sentence.

.SH OPTIONS

.BR \-c " <file>  or " \-\-config =<file>
.RS
set the configuration using 'file'.

you can use
.B -c lang/config-file
to select the 'config-file' for an installed language 'lang'
.RE

.BR \-\-debug =<module><level>,...
.RS
set debug level per module, indicated by a single letter:
Tagger (T), Tokenizer (t), Lemmatizer (l), Morphological Analyzer (a),
Chunker (c), Multi\(hyWord Units (m), Named Entity Recognition (n),
or Parser (p). Different modules must be separated by commas.

(e.g. \-\-debug=l5,n3 sets the level for the Lemmatizer to 5 and for the NER
to 3 )

Debugging lines are written to a file
.BR frog.<number>.debug
.RE
The name of that file is given at the end of the run.

.BR \-d " <level>"
.RS
set a global debug level for all modules at once.
.RE

.BR \-\-deep\(hymorph
.RS
Add a deep morphological analysis to the 'Tabbed' and JSON output.
This analysis is added to XML unconditionally.

.RE

.BR \-\-compounds
.RS
include compound information as a column to 'Tabbed' output.
This information is added to XML and JSON unconditionally.

.RE

.BR \-e " <encoding>"
.RS
set input encoding. (default UTF8)
.RE

.BR \-h " or " \-\-help
.RS
give some help
.RE

.BR \-\-language= <comma\ separated\ list\ of\ languages>
.RS
Set the languages to work on. This parameter is passed to the tokenizer.
The strings are assumed to be ISO 639\-2 codes.

The first language in the list will be the default, unspecified languages are
asumed to be of that default.

e.g. \-\-language=nld,eng,por
means: detect Dutch, English and Portuguese, with Dutch being the default,
using TextCat. Mainly useful for XML processing.

Specifying a unsupported language is a fatal error. However, you can add the
special language 'und' which assures that sentences in an unknown languages
will be labeled as such, and processed no further.

.B IMPORTANT
Frog can at the moment handle only one language at a time, as determined by the
config file. So other languages mentioned here will be tokenized correctly, but
further they will be handled as that language.
.RE

.BR \-n
.RS
assume inputfile to have one sentence per line. (newline separators)

Very useful when running interactive, otherwise an empty line is needed to
signal end of input.
.RE

.BR \-\-nostdout
.RS
suppress the 'Tabbed' or JSON output to stdout. (when no outputfile was
specified with \-o or \-\-outputdir)

Especially useful when XML output is specified with \-X or \-\-xmldir.
.RE


.BR \-o " <file>"
.RS
send 'Tabbed' output to 'file' instead of stdout. Defaults to the name of the
inputfile with '.out' appended.
.RE

.BR \-\-outputdir " <dir>"
.RS
send all 'Tabbed' or JSON output to 'dir' instead of stdout. Creates filenames
from the inputfilename(s) with '.out' appended.
.RE

.BR \-\-retry
.RS
assume a re-run on the same input file(s). Frog wil only process those files
that haven't been processed yet.
.RE


.BR \-\-skip =[tlacnmp]
.RS
skip parts of the process: Tokenizer (t), Lemmatizer (l), Morphological
Analyzer (a), Chunker (c), Named Entity Recognition (n), Multi-Word Units (m)
or Parser (p).

The Tagger cannot be skipped.

Skipping the Multiword Unit implies disabling the Parser too.
.RE

.BR \-\-alpino
.RS
Use a locally installed Alpino parser. Disables our build-in Dependency parser
.RE

.BR \-\-alpino =server
.RS
use a remote installed Alpino server, as specified in the frog configuration
file.
.RE

\" .BR \-Q
\" .RS
\" Enable quotedetection in the tokenizer. NOT USED.
\" .RE

.BR \-S " <port>"
.RS
Run Frog as a server on 'port'
.RE

.BR \-t " <file>"
.RS
process 'file'.

This option can be omitted. Frog will run on any <file> found on the
qcommand-line.
Wildcards are allowed too. When NO files are specified, Frog will start in
interactive mode.

Files with the extension '.gz' or '.bz2' are handled too. The corresponding
output-files will be compressed using the same compression again. Except
when an explicit output filename is specified.
.RE

.BR \-x " <xmlfile>"
.RS
process 'xmlfile', which is supposed to be in FoLiA format! If 'xmlfile' is
empty, and
.BR \-\-testdir =<dir>
is provided, all '.xml' files in 'dir' will be processed as FoLia XML.

This option can be omitted. Frog will process files with the 'xml' extension
as FoLiA files.

Files with the extension '.xml.gz' or '.xml.bz2' are handled too. The
corresponding output-files will be compressed using the same compression again.
Except when an explicit output filename is specified.
.RE

.BR \-X " <xmlfile>"
.RS
When 'xmlfile' is specified, create a FoLiA XML output file with that name.

When 'xmlfile' is empty, generate FoLiA XML output for every inputfile.
.RE

.BR \-\-textclass "=<cls>"
.RS
When
.BR \-x
is given, use 'cls' to find AND store text in the FoLiA document(s).
Using \-\-inputclass and \-\-\outputclass is in general a better choice.
.RE

.BR \-\-inputclass "=<cls>"
.RS
use 'cls' to find text in the FoLiA input document(s).
.RE

.BR \-\-outputclass "=<cls>"
.RS
use 'cls' to output text in the FoLiA input document(s).
Preferably this is another class then the inputclass.
.RE

.BR \-\-testdir =<dir>
.RS
process all files in 'dir'. When the input mode is XML, only '.xml' files,
'.xml.gz' or '.xml.bz2' files are taken from 'dir'. see also
.B \-\-outputdir
.RE

.BR \-\-uttmarker =<mark>
.RS
assume all utterances are separated by 'mark'. (the default is none).
.RE

.BR \-\-threads =<n>
.RS
use a maximum of 'n' threads. The default is to take whatever is needed.
In servermode we always run on 1 thread per session.
.RE

.BR \-V " or " \-\-version
.RS
show version info
.RE

.BR \-\-xmldir =<dir>
.RS
generate FoLiA XML output and send it to 'dir'. Creates filenames from the
inputfilename with '.xml' appended. (Except when it already ends with '.xml')
.RE

.BR \-X " <file>"
.RS
generate FoLiA XML output and send it to 'file'. Defaults to the name of the
inputfile(s) with '.xml' appended. (Except when it already ends with '.xml')
.RE

.BR \-\-id "=<id>"
.RS
When
.BR \-X
for FoLia is given, use 'id' to give the doc an ID. The default is an xml:id
based on the filename.
.RE

.BR \-\-allow\-word\-corrections
.RS
Allow the
.BR ucto
tokenizer to apply simple corrections on words while processing FoLiA output.
For instance splitting punctuation.
.RE

.BR \-\-max\-parser\-tokens "=<num>"
.RS
Limit the size of sentences to be handled by the Parser. (Default 500 words).

The Parser is very memory consuming. 500 Words will already need 16Gb of RAM.
.RE

.BR \-\-JSONin
.RS
The input is in JSON format. Mainly for Server mode, but works on files too.

This implies \-\-JSONout too!
.RE

.BR \-\-JSONout
.RS
Output will be in JSON instead of 'Tabbed'.
.RE

.BR \-\-JSONout "=<indent>"
.RS
Output will be in JSON instead of 'Tabbed'. The JSON will be idented by value
 'indent'. (Default is indent=0. Meaning al the JSON will be on 1 line)
.RE

.BR \-T
or
.BR \-\-textredundancy "=[full|medium|none]"
.RS
Set the text redundancy level in the tokenizer for text nodes in FoLiA output:
.B full
add text to all levels: <p> <s> <w> etc.
.B minimal
don't introduce text on higher levels, but retain what is already there.
.B none
only introduce text on <w>, AND remove all text from higher levels
.RE

.BR \-\-override "=<section>.<parameter>=<value>"
.RS
Override a parameter from the configuration file with a different value.

This option may be repeated several times.
.RE

.SH BUGS
likely

.SH AUTHORS
Maarten van Gompel

Ko van der Sloot

Antal van den Bosch

e\-mail: lamasoftware@science.ru.nl
.SH SEE ALSO
.BR ucto (1)
.BR mblem (1)
.BR mbma (1)
.BR ner (1)


NEWS
0.28 - 2023-02-28
[Ko van der Sloot]
*  We no longer accept FoLiA paragraphs with both Words and Sentences.

[Maarten van Gompel]
* Software metadata fix

0.27.2 - 2023-02-23
[Maarten van Gompel]
* Minor software metadata fix only, no functional changes

0.27.1 - 2023-02-22
[Maarten van Gompel]
* Software metadata update only, no functional changes

0.27 - 2023-02-22
[Ko van der Sloot]
* Major Release.
* Internally we always perform a 'deep' morphological analysis.
* This information is used for XML and JSON output.
* For the 'classic' Tabbed output, we maintain backward comptability.
* You need to specify '--deep-morph' to get the deep analysis in the output.
* You may also specify '--compounds' to get an extra column with compound
  information.

Other changes:
* C++ code quality
* adapted to more recent Timbl implementations (Unicode awareness)
* Tokenizer:
  - Better handling of --languages option.
  - 'und' is now also acceptable as a "language"
  - Better debugging possibility
* Mbma: To many alternatives with Inverted Verbs were generated. As the
        Tagger doesn't help us directly, we filter on the person of the next
	word, and only return V/te2I when the next word is 2-nd person

0.26 - 2023-01-02
[Ko van der Sloot]
* fix for https://github.com/LanguageMachines/frog/issues/96
* code improvements, readability and fixing CppCheck warnings
* needs recent ticcutils (>=0.30)
* needs newest Timbl (6.8) for more Unicode awarenes
* updated GigHub action

[Maarten van Gompel]
* added MAINTAINERS file
* updated codemeta.json

0.25 - 2022-07-22

[Maarten van Gompel]
* updated metadata (codemeta.json) following new (proposed) CLARIAH requirements (CLARIAH/clariah-plus#38)
* added builds-deps.sh for automatically building and installing dependencies
* added Dockerfile and instructions
* added support for user-based configuration dirs ($XDG_CONFIG_HOME/frog), takes precedence over global data dirs

[Ko vd Sloot]
* updated Doxygen config file

0.24 - 2021-12-15
[Ko vd Sloot]
* start using the newest UTF8 aware Timbl and Mbt and Ucto
* use NFC normalized UnicodeString more general internaly
* added a fix in MBMA codng, to get better reproducable result on different
  OS/Compiler combinations
* lots of small refactoring
* bumped library version, because of some API changed

[Maarten van Gompel]
* merged a patch suggested by Helmut Grohne <helmut@subdivi.de>
  - configure.ac: Bug#993123: frog FTCBFS: hard codes the build
    architecture pkg-config Source: frog Version: 0.20-2
      Tags: patch upstream User:
    debian-cross@lists.debian.org Usertags:
      ftcbfs frog fails to cross build from source, because configure.ac hard
      codes the build architecture pkg-config in one place (after
      correctly detecting the host architecture one). Simply using the
      correct substitution variable makes frog cross buildable. Please
      consider applying the attached patch.
        Helmut Signed-off-by: Maarten van Gompel <proycon@anaproy.nl>

0.23 - 2021
* requires libfolia 2.9
* replaced TravisCI by GitHub actions
* fixed https://github.com/proycon/python-frog/issues/20
* some refactoring to avoid unneeded creation of files

0.22 - 2020-11-17
[Ko vd Sloot]
* start using the tmp_stream() class form ticcutils 0.25

0.21 - 2020-07-22
[Ko van der Sloot]
* Fixes a problem with temporary files not being cleaned up properly #92

0.20.1 - 2020-04-15
[Ko vd Sloot]
Bug fix release.
- added missing Doxygen.cfg to the tarball

0.20 - 2020-04-15
[Ko vd Sloot]
* added Doxygen to the build
* added a lot of comment in Doxygen format
* adapted to the newest ticcutils version
* adapted to latest libfolia
* adapted to latest ucto
* lots of code refactorings
* implemented --JSONin option (server only)
* implemented --JSONout option
* added a --allow-word-correction option which allows ucto to correct FoLiA
  Word nodes

[Iris Hendrix]
Documentation updates

0.19.1 - 2019-11-15
[Ko vd Sloot]
* fixed an overseen incompatability problem with the new libfolia.
   (https://github.com/proycon/tscan/issues/13)
* removed dependency on MbtServer
* Some documentation updates
* improved using Alpino, using unique filenames now.

0.19 - 2019-10-21
[Ko vd Sloot]
* added code to use al locally installed Alpino parser
* added code to use a remote Alpino Server
* added code to use (remote) timblservers and mbtservers for alle modules
  using JSON calls. Stil experimental.
* several code refactoring and small fixes:
  - memory leaks
  - using NER files in non-standard locations
  - bug fixes for some corner cases.

* frog.*.debug files  are cleaned up after 1 day.

0.18.3 - 2019-07-22
[Ko vd Sloot]
Bug fix release:
* Fixes:
  - https://github.com/LanguageMachines/frog/issues/78

0.18.2 - 2019-07-15
[Ko vd Sloot]
Bug fix release:
* Fixes for:
  - https://github.com/LanguageMachines/frog/issues/75
  - https://github.com/LanguageMachines/frog/issues/77

0.18.1 - 2019-06-19
[Ko vd Sloot]
Bug fix release:
"tabbed" output contained 1 tab to much when --skip=a was specified

0.18 - 2019-06-19
[Ko vd Sloot]
Bug fixes and enhancements:
* provenance uses new 'generate_id' option in libfolia:processor
* solved problems when frogging partly tokenized FoLiA
* solved problems when processing with --skip=t
* small improvement in compound detection (still more to do...)

0.17 - 2019-05-29
[Ko vd Sloot]
This release supports FoLiA 2.0
* some bug fixes
  - trust the tokenizer to get the default language
  - don't stumble upon empty sentences introduced by a non-breaking-space
  - provenance data is added for all the modules

0.16 - 2019-05-15
[Ko vd Sloot]
This is the last release using pre FoLiA 2.0
It includes a total rework of the Frog Internals, aiming at better
maintainability and hoping for a speedup and a smaller memory footprint.
This work will continue in the upcoming release for FoLiA 2.0

Major Changes:
* total rework. Not using a FoLiA document as the internal datastructure anymore
  but a FrogData structure.

src/Frog.cxx
/* ex: set tabstop=8 expandtab: */
/*
  Copyright (c) 2006 - 2023
  CLST  - Radboud University
  ILK   - Tilburg University

  This file is part of frog:

  A Tagger-Lemmatizer-Morphological-Analyzer-Dependency-Parser for
  several languages

  frog is free software; you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation; either version 3 of the License, or
  (at your option) any later version.

  frog is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with this program.  If not, see <http://www.gnu.org/licenses/>.

  For questions and suggestions, see:
      https://github.com/LanguageMachines/frog/issues
  or send mail to:
      lamasoftware (at ) science.ru.nl
*/

#include <cstdlib>
#include <cstdio>
#include <sys/wait.h>
#include <signal.h>
#include <string>
#include <cstring>
#include <random>
#include <iostream>
#include <sstream>
#include <fstream>
#include <vector>
#include "config.h"

#include "timbl/TimblAPI.h"
#include "ticcutils/FdStream.h"
#include "ticcutils/ServerBase.h"

// individual module headers

#include "frog/Frog-util.h"
#include "frog/FrogAPI.h" //public API interface
#include "frog/ucto_tokenizer_mod.h"
#include "frog/tagger_base.h"
#include "ticcutils/StringOps.h"
#include "ticcutils/CommandLine.h"
#include "ticcutils/FileUtils.h"

using namespace std;

#define LOG *TiCC::Log(theErrLog)
#define DBG *TiCC::Log(theDbgLog)

/* assumptions:
   each components gets its own configfile per cmdline options
   another possibility is the U: en K: style option seperation

   each component has a read_settings_file in own namespace,
        which sets variables global to that namespace
	to be used in the init() for that namespace

   Further,
   each component provides a Test(File) which writes output to
        File.<componentname>_out
   and a Classify(Instance) which produces relevant output
        as a string return
	or somehwere else,
   to be determined later,
   after pre- and postprocessing raw classification data

*/

void usage( ) {
  cout << endl << "Options:\n";
  cout << "\t========================== INPUT options ==================================\n"
       << "\t -e <encoding>          specify encoding of the input (default UTF-8)\n"
       << "\t -t <testfile>          Run frog on this file\n"
       << "\t -x <testfile>          Run frog on this FoLiA XML file. Or the files from 'testdir'\n"
       << "\t --textclass=<cls>      use the specified class to search for text in the the FoLiA docs. (default 'current'). \n"
       << "\t                        the same value is used for output too.\n"
       << "\t                        Deprecated! use --inputclass and --outputclass.\n"
       << "\t --inputclass=<cls>     use the specified class to search for text in the the FoLiA docs. (default 'current') \n"
       << "\t --outputclass=<cls>    use the specified class to output text in the the FoLia docs. (default 'inputclass') \n"
       << "\t --testdir=<directory>  All files in this dir will be tested\n"
       << "\t --uttmarker=<mark>     utterances are separated by 'mark' symbols (default none)\n"
       << "\t -n                     Assume input file to hold one sentence per line\n"
       << "\t --retry                assume frog is running again on the same input,\n"
       << "\t                        already done files are skipped. (detected on the basis of already existing output files)\n"
       << "\t --max-parser-tokens=<n> inhibit parsing when a sentence contains over 'n' tokens. (default: 500, needs already 16Gb of memory!)\n"
    //       << "\t -Q                     Enable quote detection in tokenizer.\n"
       << "\t --JSONin               The input is JSON. Implies JSONout too! (server mode only)\n"
       << "\t -T or --textredundancy=[full|minimal|none]\n"
       << "\t                        Set the text redundancy level in the tokenizer for text nodes in FoLiA output: \n"
       << "\t                        'full' - add text to all levels: <p> <s> <w> etc.\n"
       << "\t                        'minimal' - don't introduce text on higher levels, but retain what is already there.\n"
       << "\t                        'none' - only introduce text on <w>, AND remove all text from higher levels\n"
       << "\t ============= CONFIGURATION OPTIONS =====================================\n"
       << "\t -c <filename> OR --config=<filename>\n"
       << "\t                        use this configuration file (default " << FrogAPI::defaultConfigFile("nld") << ")\n"
       << "\t                        you can use -c lang/config-file to select the config-file for an installed language 'lang'\n"
       << "\t --language <language-list>\n"
       << "\t                        Set the language(s) to be used in the tokenizer.\n"
       << "\t                        e.g. --language=nld,eng,por\n"
       << "\t                        The first language in the list will be the default. (default dutch).\n"
       << "\t                        IMPORTANT: frog can handle only one language at a time, determined by the config.\n"
       << "\t                        So other languages mentioned here will be tokenized correctly, but further handled as the configured language.\n"
       << "\t --override <section>.<parameter>=<value>\n"
       << "\t                        Override a configuration option, can be used multiple times\n"
       << "\t ============= OUTPUT OPTIONS ============================================\n"
       << "\t -o <outputfile>        Output Tab separated or JSON output to file, instead of default stdout\n"
       << "\t --nostdout             suppress Tabbed/JSON output to stdout\n"
       << "\t -X <xmlfile>           Output also to an XML file in FoLiA format\n"
       << "\t --id=<docid>           Document ID, used in FoLiA output. (Default 'untitled')\n"
       << "\t --allow-word-corrections         allow the tokenizer to correct <w> nodes. (FoLiA only)\n"
       << "\t --outputdir=<dir>      Output to dir, instead of default stdout\n"
       << "\t --xmldir=<dir>         Use 'dir' to output FoliA XML to.\n"
       << "\t --deep-morph           add deep morphological information to the output\n"
       << "\t                        (default for XML)\n"
       << "\t --compounds            add compound information to the output\n"
       << "\t                        (default for XML and JSON)\n"
       << "\t --JSONout=n            Output JSON instead of Tabbed.\n"
       << "\t                        When n != 0, use it for pretty-printing the output. (default n=0) \n"
       << "\t ============= MODULE SELECTION ==========================================\n"
       << "\t --skip=[mptncla]       Skip Tokenizer (t), Lemmatizer (l), Morphological Analyzer (a), Chunker (c),\n"
       << "\t                        Multi-Word Units (m), Named Entity Recognition (n), or Parser (p)\n"
       << "\t --alpino               use a locally installed Alpino parser\n"
       << "\t --alpino=server        use a remote installed Alpino server\n"
       << "\t                        (as specified in the frog configuration file)\n"
       << "\t ============= OTHER OPTIONS ============================================\n"
       << "\t -h or --help           give some help.\n"
       << "\t -V or --version        Show version info.\n"
       << "\t -d <debug level>  (for more verbosity)\n"
       << "\t --debug=<module><level>,<module><level>... (eg --debug=l5,n3) \n"
       << "\t                        Set debug value for Tagger (T), Tokenizer (t), Lemmatizer (l), Morphological Analyzer (a), Chunker (c),\n"
       << "\t                        Multi-Word Units (m), Named Entity Recognition (n), or Parser (p)\n"
       << "\t -S <port>              Run as server instead of reading from testfile, using 'port' \n"
#ifdef HAVE_OPENMP
       << "\t --threads=<n>          Use a maximum of 'n' threads. Default: 8. \n"
#endif
       << "\t                         (but always 1 for server mode)\n";
}


unsigned long long random64(){
  /// generate a 64 bit random number
  static std::random_device rd;
  static std::uniform_int_distribution<unsigned long long> dis;
  static std::mt19937_64 gen(rd());
  return dis(gen);
}

string randnum( int len ){
  /// generate a rather random string of length \e len
  /*!
    \param len The length of the output string

    We generate a 64 bit random number an convert is to a string.
    Then this string is truncated at \em len characters. (which means that
    the result migth be NOT unique!)
  */
  stringstream ss;
  ss << random64() << endl;
  string result = ss.str();
  return result.substr(0,len);
}

int main(int argc, char *argv[]) {
  /// Frog's main program.
  cerr << "frog " << VERSION << " (c) CLST, ILK 1998 - 2023" << endl
       << "CLST  - Centre for Language and Speech Technology,"
       << "Radboud University" << endl
       << "ILK   - Induction of Linguistic Knowledge Research Group,"
       << "Tilburg University" << endl;
  cerr << "based on [" << Tokenizer::VersionName() << ", "
       << folia::VersionName() << ", "
       << Timbl::VersionName() << ", "
       << TiCCServer::VersionName() << ", "
       << Tagger::VersionName() << "]" << endl;
  TiCC::LogStream *theErrLog = 0;
  ostream *the_dbg_stream = 0;
  TiCC::LogStream *theDbgLog = 0;
  std::ios_base::sync_with_stdio(false);
  FrogOptions options;
  string db_filename;
  try {
    TiCC::CL_Options Opts("c:e:o:t:T:x::X::nQhVd:S:",
			  "config:,testdir:,"
			  "help,textclass:,inputclass:,outputclass:,"
			  "uttmarker:,max-parser-tokens:,textredundancy:,"
			  "skip:,id:,outputdir:,xmldir:,tmpdir:,deep-morph,"
			  "compounds,language:,retry,nostdout,ner-override:,"
			  "debug:,keep-parser-files,version,threads:,alpino::,"
			  "override:,KANON,TESTAPI,debugfile:,JSONin,JSONout::,"
			  "allow-word-corrections,OLDMWU");
    Opts.init(argc, argv);
    if ( Opts.is_present('V' ) || Opts.is_present("version" ) ){
      // we already did show what we wanted.
      return EXIT_SUCCESS;
    }
    if ( Opts.is_present( 'h' )
	 || Opts.is_present( "help" ) ) {
      usage();
      return EXIT_SUCCESS;
    };
    string remove_command = "find frog.*.debug -mtime +1 -exec rm {} \\;";
    cerr << "removing old debug files using: '" << remove_command
	 << "'" << endl;
    if ( system( remove_command.c_str() ) ){
      // nothing
    }
    theErrLog = new TiCC::LogStream( cerr, "frog-", StampMessage );
    Opts.extract( "debugfile", db_filename );
    if ( db_filename.empty() ){
      db_filename = "frog." + randnum(8) + ".debug";
    }
    the_dbg_stream = new ofstream( db_filename );
    theDbgLog = new TiCC::LogStream( *the_dbg_stream, "frog-", StampMessage );
    FrogAPI frog( Opts, theErrLog, theDbgLog );
    if ( !frog.options.fileNames.empty() ) {
      frog.run_on_files();
    }
    else if ( frog.options.doServer ) {
      frog.run_a_server();
    }
    else {
      // interactive mode
      frog.run_interactive();
    }
  }
  catch ( const TiCC::OptionError& e ){
    usage();
    return EXIT_FAILURE;
  }
  catch ( const exception& e ){
    LOG << "fatal error: " << e.what() << endl;
    return EXIT_FAILURE;
  }
  delete theErrLog;
  delete theDbgLog;
  delete the_dbg_stream;

  if ( !db_filename.empty() ){
    ifstream test(db_filename,ifstream::ate);
    if ( test.tellg() > 0 ){
      cerr << "Some debugging information is available in: "
	   << db_filename << endl;
    }
    else {
      remove( db_filename.c_str() );
    }
  }
  return EXIT_SUCCESS;
}
